pkg/kubelet/dockertools/docker.go

```diff
@@ -691,3 +693,63 @@ type ContainerCommandRunner interface {
 	RunInContainer(containerID string, cmd []string) ([]byte, error)
 	GetDockerServerVersion() ([]uint, error)
 }
+
+type DockerCache interface {
+	KubeletRunningDockerContainers() (DockerContainers, error)
+}
+
+func NewDockerCache(client DockerInterface) DockerCache {
+	containers, err := GetKubeletDockerContainers(client, false)
+	cacheTime := time.Now()
+	return &dockerCache{
+		client:        client,
+		cacheTime:     cacheTime,
+		containers:    containers,
+		err:           err,
+		updatingCache: false,
+	}
+}
+
+// dockerCache is a default implementation of DockerCache interface
+type dockerCache struct {
+	client        DockerInterface
+	lock          sync.RWMutex
+	cacheTime     time.Time
+	containers    DockerContainers
+	err           error
+	updatingCache bool
+	lastReqTime   time.Time
+}
+
+func (d *dockerCache) KubeletRunningDockerContainers() (DockerContainers, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	if time.Since(d.cacheTime) > 2*time.Second {
+		d.containers, d.err = GetKubeletDockerContainers(d.client, false)
+		d.cacheTime = time.Now()
+	}
+	d.lastReqTime = time.Now()
+	if !d.updatingCache {
+		d.updatingCache = true
+		go d.startUpdatingCache()
+	}
+	return d.containers, d.err
+}
+
+func (d *dockerCache) startUpdatingCache() {
+	run := true
+	for run {
+		_ = <-time.After(100 * time.Millisecond)
+		containers, err := GetKubeletDockerContainers(d.client, false)
```

Comment:
WDYT of treating an error as not having something in the cache? It will give us a chance to retry on the request and won't make us have to share an error